config.ini.txt
; config.ini is case-sensitive
; Lists should be seperated with commas
; Boolean values should be "true" or "false" (anything other than exactly "true" is considered false; this includes "True" or "1")
; Almost all parameters are optional.



[core]      ; Sets core settings (eg. relating to how input / output is processed)
timeout=3                                 ; Default: 3. After this many seconds without communication from the printer, the "timeout" function of all peripherals will be executed.
include_channel_count_in_status=true      ; Default: true. If true, response to F13 commands will add a "channel_count" parameter indicating how many channels the IFS Jacker can provide
include_peripherals_in_status=true        ; Default: (none). If true, response to F13 commands will include the status of all peripherals. If false, F13 will not include peripheral statuses. Overrides the next param if present.
peripherals_in_status=-1                  ; Default: -1. If negative, F13 responses will include the status of all peripherals. If 0, F13 will not include peripheral statuses. If positive, F13 will include (up to) that many peripheral statuses each time, rotating between them.
mmu_timeout=0.065                         ; Default: 0.065. Sets how long IFS Jacker will wait for an IFS to respond to a command before timing out. (Giving up, and in the case of an F13 request, assuming all channels empty and idle.)
mmu_idle_command_delay=0.01               ; Default: 0.01. Sets how long IFS Jacker will wait after receiving an MMU response, before issuing the next idle F13 command. Ignored for any non-idle command. Peripherals may use this time to execute "slow" commands, knowing that no MMU response is being awaited.
force_present_channels=1,2                ; Default: (none). Any channels listed here will always be reported as having filament in them regardless of what the IFS says. 1-based.
force_absent_channels=3,4                 ; Default: (none). Channels listed here will always be reported empty. If a channel is listed in both of these, absent has priority.



[console]      ; Sets settings relating to the console input / output
exclude_categories=silent      ; Default: silent. Any category of data listed here will not be output to the console
include_categories=error,info  ; Default: (none). If this is present and not empty, ONLY categories listed here (with a few exceptions during startup / shutdown) will be output to console

; Valid categories:
; config - Output from loading configuration data
; data - Commands being sent / received between the printer and IFSes
; debug - Generally not used in release versions; used for debug testing info
; error - Errors / exceptions
; info - General info



[uart_multi_X]      ; Preconfiguration for any UART instances which have multiple devices connected to them. Replace X with a name or number to identify the connection.
uart_instance=1      ; Must be specified. Which hardware UART instance to use. PIO / SoftwareSerial are not currently supported (and SoftwareSerial probably never will be, but PIO is a goal).
tx_pin=4             ; Must be specified.
rx_pin=5             ; Must be specified.
en_write_state=true  ; Default: true. Whether to set EN pins high (true) or low (false) to enable writing. For most RS485 boards this will be high.
baud=115200          ; Default: 115200.
bits=8               ; Default: 8.
parity=parity        ; Default: None. You can specify the literal word "None" if you want to explicitly use none.
stop_bits=1          ; Default: 1.



[printer]      ; Printer settings. Omit this altogether if you don't want an upstream printer connection (eg. for testing, or for custom integrations over USB with other printers).
type=uart_en      ; Must be specified. Connection type. Valid values are 'uart' (pure UART) or 'uart_en' (UART with EN pin).
uart_instance=0      ; Must be specified. Which hardware UART instance to use.
tx_pin=0             ; Must be specified.
rx_pin=1             ; Must be specified.
en_pin=2             ; Must be specified if type is "uart_en". Not used otherwise.
en_write_state=true  ; Default: true. Only used if type is "uart_en". Whether to set EN pins high (true) or low (false) to enable writing. For most RS485 boards this will be high.
baud=115200          ; Default: 115200.
bits=8               ; Default: 8.
parity=parity        ; Default: None. You can specify the literal word "None" if you want to explicitly use none.
stop_bits=1          ; Default: 1.



[mmu_0]      ; IFS settings for the first IFS. And mmu_1 for the second, mmu_2 for the third, and so on. This example section is for an MMU on a dedicated UART.
type=uart_en         ; Must be specified (unless a multi_id is specified instead; see below). Connection type. Valid values are 'uart' (pure UART) or 'uart_en' (UART with EN pin).
uart_instance=1      ; Must be specified. Which hardware UART instance to use.
tx_pin=4             ; Must be specified.
rx_pin=5             ; Must be specified.
en_pin=6             ; Must be specified if type is "uart_en". Not used otherwise.
en_write_state=true  ; Default: true. Only used if type is "uart_en". Whether to set EN pins high (true) or low (false) to enable writing. For most RS485 boards this will be high.
baud=115200          ; Default: 115200.
bits=8               ; Default: 8.
parity=parity        ; Default: None. You can specify the literal word "None" if you want to explicitly use none.
stop_bits=1          ; Default: 1.

[mmu_1]      ; This example section is for an MMU connected to a multiple-IFS-on-one-UART setup.
multi_id=X      ; Must be specified (unless a type is specified instead, see above). Replace X with the name of your multi setup. For example if you used [uart_multi_Kevin], you would put "multi_id=Kevin".
en_pin=7        ; Must be specified.



[peripheral_0]      ; Defines a peripheral (an added input/output of some kind, which may be purely software, that can be accessed via the IFS Jacker)
type=dummy      ; Must be specified. Type of peripheral. This should match the script and class name (p_dummy.py / IJP_Dummy) but always lowercase.
identifier=Sample Peripheral      ; Default depends on type. Specifies a name for the peripheral (shown in Z3 response).
report_f13=true      ; Default depends on type. Whether or not to show the peripheral's status code in F13 responses.
report_z4=true      ; Default is whatever report_f13 is set to (which may be its default). Same as above but for Z4.
; Additional parameters will exist depending on the specific peripheral type.

src/config.py
from util import load_ini_file
from console import get_console
from core import IJ_Core
from p_dummy import IJP_Dummy
from peripheral import load_peripheral
import comm, gc

STANDARD_COMM_GENERATION_TYPES = [comm.IJ_Comm_UART, comm.IJ_Comm_UART_EN]

def load_config(core: IJ_Core, comm_mgr: comm.IJ_Comm_Manager):
    ini_data = load_ini_file('config.ini')
    console = get_console()

    core_data = ini_data.get('core', {})
    core.printer_connected_timeout = int(float(core_data.get('timeout', 3)) * 1000)
    core.include_channel_count_in_status = core_data.get('include_channel_count_in_status', 'true') == 'true'
    old_include_peripherals_in_status = core_data.get('include_peripherals_in_status', None)
    if old_include_peripherals_in_status is not None:
        if old_include_peripherals_in_status == 'true':
            core.peripherals_in_status_count = -1
        else:
            core.peripherals_in_status_count = 0
    else:
        core.peripherals_in_status_count = int(core_data.get('peripherals_in_status', -1))
    core.mmu_response_timeout = int(float(core_data.get('mmu_timeout', 0.065)) * 1000)
    core.mmu_idle_command_delay = int(float(core_data.get('mmu_idle_command_delay', 0.01)) * 1000)
    force_present_channels = [item.strip() for item in core_data.get('force_present_channels', '').split(',')]
    force_absent_channels = [item.strip() for item in core_data.get('force_absent_channels', '').split(',')]
    if len(force_absent_channels) > 1 or force_absent_channels[0] != '':
        mask = 0
        for i in [int(channel) for channel in force_absent_channels]:
            mask |= 1 << i-1
        core.force_absent_mask = ~mask
    if len(force_present_channels) > 1 or force_present_channels[0] != '':
        mask = 0
        for i in [int(channel) for channel in force_present_channels]:
            mask |= 1 << i-1
        core.force_present_mask = mask & core.force_absent_mask

    console_data = ini_data.get('console', {})
    console.exclude_categories = [item.strip() for item in console_data.get('exclude_categories', '').split(',')]
    if len(console.exclude_categories) == 1 and console.exclude_categories[0] == '':
        console.exclude_categories = ['silent']
    console.include_categories = [item.strip() for item in console_data.get('include_categories', '').split(',')]
    if len(console.include_categories) == 1 and console.include_categories[0] == '':
        console.include_categories = []

    multi_connections: dict = {}
    all_comms: list = []

    for multi_key in ini_data.keys():
        if multi_key.startswith('uart_multi_') and len(multi_key) > 11:
            new_id = multi_key[11:]
            console.print(f'Loading UART multi-connection {new_id}', 'config')
            new_multi_comm = comm.IJ_Comm_UART_EN_Multi.make_from_config(ini_data[multi_key])
            multi_connections[new_id] = new_multi_comm
            all_comms.append(new_multi_comm)

    printer_data = ini_data.get('printer', None)
    if printer_data:
        console.print(f'Loading printer', 'config')
        core.printer_comm = get_comm(printer_data, multi_connections)
        all_comms.append(core.printer_comm)

    mmu_index = 0
    while f'mmu_{mmu_index}' in ini_data.keys():
        console.print(f'Loading MMU {mmu_index}', 'config')
        mmu_data = ini_data[f'mmu_{mmu_index}']
        new_comm = get_comm(mmu_data, multi_connections)
        core.mmu_comms.append(new_comm)
        all_comms.append(new_comm)
        mmu_index += 1

    comm_mgr.comm_list = all_comms

    peripherals = load_peripherals(ini_data, all_comms)
    core.peripherals = peripherals.copy()

    for this_comm in all_comms:
        this_comm.initialize()

    for this_peripheral in peripherals:
        this_peripheral.initialize()
    gc.collect()

def get_comm(data: dict[str, str], multi_connections: dict):
    multi_id = data.get('multi_id', '')
    multi_parent = multi_connections.get(multi_id, None)
    if multi_parent:
        return multi_parent.make_child(data)
    elif multi_id != '':
        raise Exception(f'Invalid multi ID "{multi_id}"')

    comm_type = f'ij_comm_{data.get('comm', '')}'
    for standard_type in STANDARD_COMM_GENERATION_TYPES:
        if standard_type.__name__.lower() == comm_type:
            return standard_type.make_from_config(data)

    raise Exception(f'Invalid comm type "{comm_type}"')

def load_peripherals(ini_data: dict[str, dict[str, str]], all_comms: list[comm._IJ_Comm_Abstract]) -> list:
    console = get_console()

    console.print('Loading peripherals', 'config')
    highest_index = -1
    for sec_name in ini_data.keys():
        if sec_name.startswith('peripheral_'):
            try:
                this_index = int(sec_name[11:])
                highest_index = max(highest_index, this_index)
            except:
                pass

    result = []
    for i in range(highest_index + 1):
        if f'peripheral_{i}' in ini_data:
            console.print(f'Loading peripheral {i}', 'config')
            try:
                peripheral_sec = ini_data[f'peripheral_{i}']
                new_peripheral = load_peripheral(i, peripheral_sec, all_comms)

                result.append(new_peripheral)
            except KeyboardInterrupt:
                raise        
            except Exception as e:
                new_peripheral = IJP_Dummy()
                new_peripheral.identifier = "Failed to load"
                result.append(new_peripheral)
                console.print_exception(e, 'config')
        else:
            new_peripheral = IJP_Dummy()
            new_peripheral.identifier = "Placeholder dummy"
            result.append(new_peripheral)

    return result
